clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java

```diff
@@ -0,0 +1,254 @@
+ */
+
+package tachyon.client;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileChannel.MapMode;
+
+import com.google.common.primitives.Ints;
+import com.google.common.io.Closer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * <code>LocalBlockOutStream</code> implementation of TachyonFile. This class is not client facing.
+ */
+public class LocalBlockOutStream extends BlockOutStream {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final int mBlockIndex;
+  private final long mBlockCapacityByte;
+  private final long mBlockId;
+  private final long mBlockOffset;
+  private final Closer mCloser = Closer.create();
+  private final String mLocalFilePath;
+  private final RandomAccessFile mLocalFile;
+  private final FileChannel mLocalFileChannel;
+  private final ByteBuffer mBuffer;
+
+  private long mAvailableBytes = 0;
+  private long mInFileBytes = 0;
+  private long mWrittenBytes = 0;
+
+  private boolean mCanWrite = false;
+  private boolean mClosed = false;
+
+  /**
+   * @param file the file the block belongs to
+   * @param opType the OutStream's write type
+   * @param blockIndex the index of the block in the file
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  LocalBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, TachyonConf tachyonConf)
+      throws IOException {
+    this(file, opType, blockIndex, tachyonConf.getBytes(Constants.USER_QUOTA_UNIT_BYTES,
+        8 * Constants.MB), tachyonConf);
+  }
+
+  /**
+   * @param file the file the block belongs to
+   * @param opType the OutStream's write type
+   * @param blockIndex the index of the block in the file
+   * @param initialBytes the initial size bytes that will be allocated to the block
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  LocalBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initialBytes,
+                      TachyonConf tachyonConf) throws IOException {
+    super(file, opType, tachyonConf);
+
+    if (!opType.isCache()) {
+      throw new IOException("LocalBlockOutStream only support WriteType.CACHE");
+    }
+
+    mBlockIndex = blockIndex;
+    mBlockCapacityByte = mFile.getBlockSizeByte();
+    mBlockId = mFile.getBlockId(mBlockIndex);
+    mBlockOffset = mBlockCapacityByte * blockIndex;
+
+    mCanWrite = true;
+
+    if (!mTachyonFS.hasLocalWorker()) {
+      mCanWrite = false;
+      String msg = "The machine does not have any local worker.";
+      throw new IOException(msg);
+    }
+    mLocalFilePath = mTachyonFS.getLocalBlockTemporaryPath(mBlockId, initialBytes);
```

Comment:
Better leave a TODO and state that LocalFileBlockWriter will be used instead of using java file API directly?